pipit/util/test_generator.py

```diff
@@ -0,0 +1,366 @@
+import random
+import textwrap
+import pandas as pd
+
+
+class FakeNode:
+    """
+    A single node of the calling tree used to produce fake traces.
+    Represents a single function call.
+    """
+
+    def __init__(
+        self, name, exc_time, mpi_type="", mpi_tgt=0, mpi_volume=0, mpi_time=0
+    ):
+        self.name = name
+        self.exc_time = exc_time
+        self.inc_time = exc_time
+        self.children = {}  # run_time -> child node
+        self.is_mpi = mpi_type != ""
+        self.mpi_type = mpi_type
+        self.mpi_tgt = mpi_tgt
+        self.mpi_volume = mpi_volume
+        self.mpi_time = mpi_time
+        self.total_nodes = 1
+        self.parent = None
+
+    def grow_inc_time(self, time):
+        """
+        This function adjusts inclusive time metric when a new child is added,
+        adding the time to each parent's inc_time, up to the root.
+        """
+        self.inc_time += time
+        if self.parent is not None:
+            self.parent.grow_inc_time(time)
+
+    def grow_total_nodes(self, amt):
+        """
+        This function adjusts the total count of nodes when a new node is added,
+        adding the new count to each parent's total_nodes, up to the root.
+        """
+        self.total_nodes += amt
+        if self.parent is not None:
+            self.parent.grow_total_nodes(amt)
+
+    def add_child(self, child, run_time):
+        """
+        This function adds a child that executes after run_time
+        exclusive time within the function represented by the current node.
+        """
+        assert run_time not in self.children
```

Comment:
Question about `run_time`: If we have a child that executes right after the other one (start time is same as end time of previous child), is the `run_time` supposed to be the same for both then because you mention it is exclusive time?

Reply:
I wanted to avoid having any `run_time`s being _exactly_ the same, since many things are ordered by time and that would lead to sorting order of the lines in the file and in the dataframe being defined by other factors. In reality, one child wouldn't run _immediately_ after the other, there would be some very small amount of exclusive overhead. In that case, the two `run_time`s would be very similar.

Reply:
Good point. I agree, we shouldn't consider that situation.